### PR TITLE
Update Settings and main script

### PR DESCRIPTION
**Pull request by Lars-DHondt-KUL**
## 1) Changes to existing settings
| Description | Motivation and Context |
|--------|-------|
| Added default value for `S.solver.run_as_batch_job` (*false*) | This was a required setting because it was used in main (i.e. before getDefaultSettings was called). This setting is now initialized to its default value in InitializeSettings |
| Changed default value for `S.OpenSimADOptions.verbose_mode` (*false*) | The verbose mode is only useful for debugging OpenSimAD errors. |
| Changed default value for `S.solver.CasADi_path` to use `casadi.GlobalOptions.getCasadiPath()`. | If casadi is on the matlab search path, you no longer have to pass it to this setting. |
| Move `save_folder` from subject to misc. | All other paths are in misc, so this is more consistent. |
| Move `savename` and `result_filename` from post_processing to misc. | All other file names are in misc, so this is more consistent.  |
| Move `IG_selection` and `IG_selection_gaitCyclePercent` from subject to solver. | These relate to the solver, while subject settings are about tuning the neuromusculoskeletal model. |
| Changed `S.subject.v_pelvis_x_trgt` to `S.misc.forward_velocity`. | Task constraints (`gaitmotion_type`) are in misc. Changed name because: 1) we impose the velocity on the forward motion of the floating base, this does not have to be pelvis; 2) trgt (target) could mean that the velocity is in the objective ( $(v-v_{tgt})^2$ ), while this is not the case. |
| Renamed `S.weights.e_arm` to `S.weights.e_torqAct`. | This is the weight for the excitation of the torque actuators, which happen to drive the arms is the example models. |
| Removed `S.subject.spasticity` and `S.subject.muscle_coordination`. | These settings are not used. |
| Set `S.misc.main_path` in initializeSettings instead of in main. | Assigning this setting in main suggests the user can change its value, while it has to be the path to the PredSim repo. |

## 2) New settings
| Setting | Default | Motivation and Context |
|--------|-------|------|
| S.solver.constr_viol_tol_ipopt | 6 | Control the constraint violation tolerance independent of the convergence tolerance (tol_ipopt). Default chosen such that constraint violation at the solution is below 1E-6. |
| S.solver.nlpsol_options | `[]` | Makes it possible to pass [options to CasADi's nlp solver](https://web.casadi.org/api/html/d4/d89/group__nlpsol.html). |
| S.solver.ipopt_options | `[]` | Makes it possible to pass [options to ipopt](https://coin-or.github.io/Ipopt/OPTIONS.html). |
| S.weights.a_exp | 2 | Choose the exponent for the muscle activation term in the cost function. |

## 3) Cleaner main script

- Removed commented settings. A full overview of settings and defaults is in the readme. Having an overview for an arbitrary selection of settings in main might cause confusion.
- Added some code to plot a figure, instead of calling `run_this_file_to_plot_figures`. Previously, running run_this_file_to_plot_figures multiple times could give problems because the array with results was not cleared between runs. If this script does not have to be called from main, we can always clear the array.
- Moved 
```
if S.solver.run_as_batch_job
    add_pred_sim_to_batch(S,osim_path)
else
    [savename] = run_pred_sim(S,osim_path);
end
```
To a new function, `runPredSim`. This way, users always call the same function, regardless of using batch jobs or not. This also avoids the possibility of calling `add_pred_sim_to_batch` with `S.solver.run_as_batch_job = false` (this could cause conflicts).

## How Has This Been Tested?

Ran main script and compared result to example result.

## Suggested tests for reviewers

Run main script.

## To do
I will update the example results, to reflect the changes to the settings. I will do this after the changes to settings are approved.

### CODE REVIEW

File: DefaultSettings/getDefaultSettings.m [hunk 1]
```diff
@@ -237,32 +242,36 @@
...
+%% post_process
+if ~isfield(S,'post_process')
+    S.post_process = [];
 end
 
 % rerun post-processing without solving OCP
 if ~isfield(S.post_process,'rerun')
     S.post_process.rerun = 0;
 end
 if S.post_process.rerun && isempty(S.post_process.result_filename)
```
Comment (VDB-Bram) [inline DefaultSettings/getDefaultSettings.m:272]: `S.post_process.results_filename` should be `S.misc.result_filename`?

File: DefaultSettings/getDefaultSettings.m [hunk 2]
```diff
@@ -291,6 +311,11 @@
     S.solver.tol_ipopt = 4;
 end
 
+% constraint violation has be below 10^(-x) at the solution
```
Comment (VDB-Bram) [inline DefaultSettings/getDefaultSettings.m:314]:
```suggestion
% constraint violation has to be below 10^(-x) at the solution
```

File: DefaultSettings/getDefaultSettings.m [hunk 3]
```diff
@@ -317,31 +342,59 @@
 end
 
 % path to CasADi installation folder
```
Comment (VDB-Bram) [inline DefaultSettings/getDefaultSettings.m:344]:
```suggestion
```

File: DefaultSettings/getDefaultSettings.m [hunk 4]
```diff
@@ -317,31 +342,59 @@
...
+                'gait cycles, respectively.'])
+        end
+        if ((strcmp(S.misc.gaitmotion_type,'FullGaitCycle')) ...
+                && (S.solver.IG_selection_gaitCyclePercent < 100))
+            error(['Cannot use an initial guess of an incomplete gait cycle ' ...
+                'for predictive simulation of a full gait cycle. Please ' ...
+                'adjust S.misc.gaitmotion_type or initial guess file.'])
+        elseif ((strcmp(S.misc.gaitmotion_type,'HalfGaitCycle')) ...
+                && (S.solver.IG_selection_gaitCyclePercent < 50))
+            error(['Cannot use an initial guess of less than half gait gait ' ...
```
Comment (VDB-Bram) [inline DefaultSettings/getDefaultSettings.m:371]:
```suggestion
            error(['Cannot use an initial guess of less than a half gait ' ...
```

File: DefaultSettings/getDefaultSettings.m [hunk 5]
```diff
@@ -530,6 +537,11 @@
     S.weights.a = 2000; 
 end
 
+% exponent for the mmuscle activations
```
Comment (VDB-Bram) [inline DefaultSettings/getDefaultSettings.m:540]:
```suggestion
% exponent for the muscle activations
```

File: OCP/OCP_formulation.m [hunk 6]
```diff
@@ -595,6 +595,7 @@
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 % Scale cost function
 Jall_sc = sum(Jall)/dist_trav_tot;
+opti.minimize(Jall_sc);
```
Comment (VDB-Bram) [inline OCP/OCP_formulation.m:598]: not clear to me why this changed location. This is not related to changes in the settings?

File: README.md [hunk 7]
```diff
@@ -214,38 +215,48 @@ This code can automatically convert an OpenSim model to the external function us
...
-- **S.solver.max_iter**: 
-	- maximal amount of iterations after wich the solver will stop. Default is *10000* [double]
+- **S.solver.CasADi_path**:
+	- Path to CasADi installation (top folder). By default, this will use the CasADi installation that is in the matlab search path.
+- **S.solver.N_meshes**: 
+	- number of mesh intervals. Default is *50* [double] for S.misc.gaitmotion_type = HalfGaitCycle and *100* for FullGaitCycle	
+- **S.solver.run_as_batch_job**: 
+	- specify if the OCP is to be solved as a batch job. Default is *false* [bool]. Batch processing requires the [Parallel Computing Toolbox](https://nl.mathworks.com/products/parallel-computing.html).
 - **S.solver.parallel_mode**: 
 	- type of parallel computing. Default is *thread* [char].
```
Comment (VDB-Bram) [inline README.md:242]: For settings like this, wouldn't it be helpful to give the other options along with the default?

File: README.md [hunk 8]
```diff
@@ -254,7 +265,7 @@ This code can automatically convert an OpenSim model to the external function us
     - used for data-informed initial guess when `S.subject.adapt_IG_pelvis_y = 1;`
 	- S.subject.IG_pelvis_y is also used to establish bounds on vertical pelvis position.
 - **S.subject.adapt_IG_pelvis_y**: 
-	- boolean to adjust the trajectory of height of pelvis from the ground for data-informed initial guess. Default is *0*. 0 means the trajectory will not be changed. If 1, the trajectory will be changed such that the average value of the trajectory is equal to s.subject.IG_pelvis_y.
+	- boolean to adjust the trajectory of height of pelvis from the ground for data-informed initial guess. Default is *0*. 0 means the trajectory will not be changed. If 1, the trajectory will be changed such that the average value of the trajectory is equal to S.subject.IG_pelvis_y.
 - **S.subject.v_pelvis_x_trgt**: 
 	- average velocity you want the model to have, in meters per second. Default is *1.25* m/s [double]
```
Comment (VDB-Bram) [inline README.md:270]:
```suggestion
```

File: main.m [hunk 9]
```diff
@@ -13,138 +13,64 @@
 [pathRepoFolder,~,~] = fileparts(pathRepo);
 
 %% Initialize S
-pathDefaultSettings = fullfile(pathRepo,'DefaultSettings');
-addpath(pathDefaultSettings)
+addpath(fullfile(pathRepo,'DefaultSettings'))
 
 [S] = initializeSettings('Falisse_et_al_2022');
```
Comment (VDB-Bram) [inline main.m:18]: Why not do it like this?
```suggestion
% name of the subject
S.subject.name = 'Falisse_et_al_2022';
[S] = initializeSettings(S.subject.name);
```

File: runPredSim.m [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+% OUTPUT:
+%   - savename (optional) -
+%   * results of the simulation will be saved in a file with this name
+% 
+% Original author: Lars D'Hondt
+% Original date: 14/August/2024
+% --------------------------------------------------------------------------
+
+
+% Make sure casadi path is set up correctly. This needs to happen before
```
Comment (VDB-Bram) [inline runPredSim.m:23]:
```suggestion
% Make sure your casadi path is set up correctly. This needs to happen before
```

File: OCP/OCP_formulation.m (hunk 6 above)
Comment (Lars-DHondt-KUL) [reply OCP/OCP_formulation.m:598]: It was in the code section for passing options to the solver, so I moved it to the section that finalizes the cost function.

File: main.m (hunk 9 above)
Comment (Lars-DHondt-KUL) [reply main.m:18]: `[S] = initializeSettings('Falisse_et_al_2022');` can also be useful for other subjects, if they use a comparable model.

**Review by dgupta7 — COMMENTED** [review]
I ran the code. It worked. I have added a comment for the runPredSim.m file

File: main.m (hunk 9 above)
Comment (VDB-Bram) [reply main.m:18]: Would it not be best practice to always have a settings file for each subject?

Comment (miriamfebrer) [reply main.m:18]: I think that this may depend on your study / application.
I usually initialise settings with 'Falisse_et_al_2022' and I only add in the settings file for each subject those that are different from default or from 'Falisse_et_al_2022'.
I would keep it as it is now

File: DefaultSettings/getDefaultSettings.m [hunk 11]
```diff
@@ -316,32 +341,59 @@
...
-% path to CasADi installation folder
-if ~isfield(S.solver,'CasADi_path')
-    S.solver.CasADi_path = [];
-elseif ~isempty(S.solver.CasADi_path) && ~isfolder(S.solver.CasADi_path)
-     error('Unable to find the path assigned to "S.solver.CasADi_path"')
-end
 
-if isempty(S.solver.CasADi_path) && S.solver.run_as_batch_job
-    error('Running a simulation as batch job requires "S.solver.CasADi_path" to contain the CasADi installation folder')
+
```
Comment (miriamfebrer) [inline DefaultSettings/getDefaultSettings.m:345]: I like this part! :)

File: README.md [hunk 12]
```diff
@@ -219,38 +220,48 @@ This code can automatically convert an OpenSim model to the external function us
...
+	- the [ipopt constraint violation tolerance](https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_constr_viol_tol) is set to 10^(-S.solver.constr_viol_tol_ipopt). A higher number gives a more precise answer, but takes more time. Default is *6* [double]. Overwrites S.solver.ipopt_options.constr_viol_tol.
 - **S.solver.max_iter**: 
-	- maximal amount of iterations after wich the solver will stop. Default is *10000* [double]
-- **S.solver.parallel_mode**: 
-	- type of parallel computing. Default is *thread* [char].
-- **S.solver.N_threads**: 
-	- number of threads in parallel mode. Default is *4* [double]. When using batch computing, this value is overwritten with the number of threads assigned to each worker in you parallel cluster.
-- **S.solver.N_meshes**: 
-	- number of mesh intervals. Default is *50* [double] for S.misc.gaitmotion_type = HalfGaitCycle and *100* for FullGaitCycle
+	- maximal amount of iterations after wich the solver will stop. Default is *10000* [double]. Overwrites S.solver.ipopt_options.max_iter.
```
Comment (miriamfebrer) [inline README.md:261]: Wouldn't it be easier to define these settings using 'S.solver.ipopt_options' always? Maybe this overwriting can be confusing?

**Comment by miriamfebrer** [comment]
1) Changes to existing settings -> I agree with all the changes and I think that they make more sense now
2) New settings -> I also agree and I think that they are useful settings
3) Cleaner main script -> perfect!

I have added small comments to some lines.

I have not run the code, as I see that Dhruv ran it without problems.

File: runPredSim.m [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+% adding a simulation to the batch
+if ~isfield(S.solver,'CasADi_path')
+    try
+        S.solver.CasADi_path = casadi.GlobalOptions.getCasadiPath();
```
[the remark this answers is not in the document]
Comment (Lars-DHondt-KUL) [reply runPredSim.m:27]: Having multiple casadi versions on the matlab search path can cause errors, but that is always the case.

File: README.md (hunk 12 above)
Comment (Lars-DHondt-KUL) [reply README.md:261]: I kept the individual settings that we had, because those are the ipopt settings that could be relevant for an average PredSim user.

File: Tests/runTestSimulation.m [hunk 14]
```diff
@@ -75,22 +71,30 @@
...
-    S.solver.CasADi_path = options.CasADi_path;
-else
-    S.solver = rmfield(S.solver,'CasADi_path');
+% Make sure casadi path is set up correctly
+if isempty(options.CasADi_path)
+    try
+        S.solver.CasADi_path = casadi.GlobalOptions.getCasadiPath();
+    catch
+        error("Please add CasADi to the matlab search path, or pass the path " + ...
+            "to your CasADi installation (top folder) to 'CasADi_path'.")
```
Comment (VDB-Bram) [inline Tests/runTestSimulation.m:91]:
```suggestion
            "to your CasADi installation (top folder) to 'options.CasADi_path'.")
```

Comment (Lars-DHondt-KUL) [reply Tests/runTestSimulation.m:91]: The input to this function is not `options.CasADi_path = path`, it is `'CasADi_path', path`. So I'm not sure if it will be clear what `options` means to someone who calls this function without looking what's inside.

Comment (VDB-Bram) [reply Tests/runTestSimulation.m:91]: Indeed, I understood wrongly that options is no struct.

**Review by VDB-Bram — APPROVED** [review]
All comments are addressed.

Successfully ran:
1. test_PredSim.m
2. main.m
